nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java

```diff
@@ -0,0 +1,341 @@
+                mqttQueue.drainTo(newBuffer);
+                mqttQueue = newBuffer;
+            }
+
+        }
+    }
+
+    @Override
+    public Collection<ValidationResult> customValidate(ValidationContext context) {
+        final Collection<ValidationResult> results = super.customValidate(context);
+        int newSize = context.getProperty(PROP_MAX_QUEUE_SIZE).asInteger();
+        if (mqttQueue == null) {
+            mqttQueue = new LinkedBlockingQueue<>(context.getProperty(PROP_MAX_QUEUE_SIZE).asInteger());
+        }
+        int msgPending = mqttQueue.size();
+        if (msgPending > newSize) {
+            results.add(new ValidationResult.Builder()
+                    .valid(false)
+                    .subject("ConsumeMQTT Configuration")
+                    .explanation(String.format("%s (%d) is smaller than the number of messages pending (%d).",
+                            PROP_MAX_QUEUE_SIZE.getDisplayName(), newSize, msgPending))
+                    .build());
+        }
+
+        return results;
+    }
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        logger = getLogger();
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) throws IOException, ClassNotFoundException {
+        qos = context.getProperty(PROP_QOS).asInteger();
+        maxQueueSize = context.getProperty(PROP_MAX_QUEUE_SIZE).asLong();
+        topicFilter = context.getProperty(PROP_TOPIC_FILTER).getValue();
+
+        buildClient(context);
+    }
+
+    @OnUnscheduled
+    public void onUnscheduled(final ProcessContext context) {
+        try {
+            synchronized (mqttClientConnectLock) {
+                if(mqttClient != null && mqttClient.isConnected()) {
+                    mqttClient.disconnect(DISCONNECT_TIMEOUT);
+                    logger.info("Disconnected the MQTT client.");
+                }
+            }
+        } catch(MqttException me) {
+            logger.error("Failed when disconnecting the MQTT client.", me);
+        }
+    }
+
+
+    @OnStopped
+    public void onStopped(final ProcessContext context) throws IOException {
+        if(processSessionFactory != null) {
+            logger.info("Finishing processing leftover messages");
+            ProcessSession session = processSessionFactory.createSession();
+            transferQueue(session);
+        } else {
+            if (mqttQueue!= null && mqttQueue.size() > 0){
+                throw new ProcessException("Attempting to stop processor but stored ProcessSessionFactory is not set, there are messages in the MQTT Queue and it is configured to " +
+                        "finish processing the messages.");
+            }
```

Comment:
This isn't exactly like Kafka because with Kafka you have to pull the records from Kafka. MQTT pushes the messages to you so you have to have an internal queue to buffer them. I'm looking at both AbstractKafkaProcessors in pr/366 but don't see how this applies.

Why would the drain be initiated when the processor stopping (so started in the unscheduled)? There still is the chance the processor would finish draining the queue by itself in any remaining onTrigger calls. Also I'm already essentially doing an onTrigger in the stop. I'm using the same method that onTrigger uses to transfer the queue.